## Introduction

The purpose of this notebook will be to preprocess our data to get it ready for Machine Learning. 

    1. Create dummy or indicator features for categorical variables
    2. Standardize the magnitude of numeric features using a scaler
    3. Split your data into testing and training datasets

## Imports

Here we are importing the necessary libraries as well as the data to help with preprocessing and to get a basic machine learning model created for the dataset. 

In [20]:
#Import necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
from datetime import datetime
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression

#Don't display future warnings
warnings.simplefilter(action='ignore', category=FutureWarning)
sns.set()

In [64]:
#Import cleaned data
data = pd.read_csv('data/clean_data/clean_data.csv')

## View Data

Here we will make some adjustements to the dataset so we have all the right data types/values that we need. 

In [65]:
#Update the data type to Strings for Product Code as they are categories, not ints. 
data['ProductCode'] = data['ProductCode'].values.astype('str')

#product codes: 1293 (amusement devices), 3295 (water slides, public), and 3259 (go-karts).
#Update the product codes values to align with their actual meaning. 

#Create a dictionary for the product codes
product_codes = {'1293':'amusement devices (1293)', '3295': 'water slides, public (3295)', '3259':'go-karts (3259)'}

#Update the product code to the proper strings
data.replace(product_codes, inplace=True)

#Convert to Date Time object
data['Treatment_Date'] = pd.to_datetime(data['Treatment_Date'])

In [66]:
data.set_index('CPSC_Case_Number', inplace = True)

We have also set the Case Number as the index since it is a unqiue identifier of the data. Let's look at the current data types for the remaining columns. 

In [67]:
data.dtypes

Treatment_Date     datetime64[ns]
Age                         int64
Sex                        object
Narrative                  object
Diagnosis                  object
Other_Diagnosis            object
BodyPart                   object
Disposition                object
ProductCode                object
Device_category            object
Device_type                object
Location                   object
Stratum                    object
PSU                         int64
Weight                    float64
dtype: object

Here we see that everything is a string except for Age, PSU, and Weight. No information was available for what PSU and Weight actually mean so I will leave them alone for now. Let's veiw the data to make sure it looks acceptable. 

In [71]:
data.head()

Treatment_Date  Age Sex  \
CPSC_Case_Number                           
180125260            2017-12-31    3   F   
180108428            2017-12-31   10   F   
180120413            2017-12-31   14   M   
180125238            2017-12-30    2   F   
180135290            2017-12-30   17   F   

                                                          Narrative  \
CPSC_Case_Number                                                      
180125260         3 YOF JUMPING BOUNCE HOUSE W/MOM JUMPED UP LAN...   
180108428         10YOF PLAYING *** AT *** AT A PARTY AT *** LOC...   
180120413         14YOM- PT WAS PLAYING *** TODAY SWELLING TO RI...   
180125238         2 YOF JUMPING IN BOUNCE HOUSE LANDED AWKWARDLY...   
180135290         17YOF DRIVING GO CART, RAN INTO BARRIER, HIT L...   

                       Diagnosis Other_Diagnosis    BodyPart  \
CPSC_Case_Number                                               
180125260               Fracture             NaN  Leg, lower   
180108428          Dental injury             NaN       Mouth   
180120413                  Other      TENDONITIS        Knee   
180125238               Fracture             NaN  Leg, lower   
180135290         Strain, sprain             NaN        Hand   

                           Disposition               ProductCode  \
CPSC_Case_Number                                                   
180125260         Treated and released  amusement devices (1293)   
180108428         Treated and released  amusement devices (1293)   
180120413         Treated and released  amusement devices (1293)   
180125238         Treated and released  amusement devices (1293)   
180135290         Treated and released           go-karts (3259)   

                              Device_category     Device_type  \
CPSC_Case_Number                                                
180125260                         Inflatables      inflatable   
180108428         Not identified or unrelated  not identified   
180120413         Not identified or unrelated  not identified   
180125238                         Inflatables      inflatable   
180135290                            Go karts         go kart   

                           Location Stratum  PSU   Weight  
CPSC_Case_Number                                           
180125260         Sports/recreation       C   32   4.7570  
180108428         Sports/recreation       C    8   4.7570  
180120413         Sports/recreation       C   90   4.7570  
180125238         Sports/recreation       C   32   4.7570  
180135290         Sports/recreation       M   54  79.1731

## Get Dummies

Let's now convert the data into dummy variable columns so that all the categorical columns are split out as 0 or 1 as opposed to strings. We will also convert the "Treatment_Day" field into new columns for day, month, and year. Lastly, we will drop the "Narrative" field as it will not be used with the machine learning algorithms. 

In [72]:
#Not including Diagnosis as that is the first prediction. 'Diagnosis',
d_data = pd.get_dummies(data, columns= ['Sex', 'Other_Diagnosis', 'BodyPart', 
                               'Disposition', 'ProductCode', 'Device_category', 
                               'Device_type', 'Location', 'Stratum', 'PSU'], drop_first=True)

#Need to convert the treatment date to Day, Month, and Year
d_data['Treatment_Day'] = d_data['Treatment_Date'].dt.day
d_data['Month_Day'] = d_data['Treatment_Date'].dt.month
d_data['Year_Day'] = d_data['Treatment_Date'].dt.year
d_data.drop('Treatment_Date', axis=1,inplace= True)
d_data.drop('Narrative', axis=1,inplace= True)

In [73]:
d_data.head()

Age       Diagnosis   Weight  Sex_M  \
CPSC_Case_Number                                        
180125260           3        Fracture   4.7570      0   
180108428          10   Dental injury   4.7570      0   
180120413          14           Other   4.7570      1   
180125238           2        Fracture   4.7570      0   
180135290          17  Strain, sprain  79.1731      0   

                  Other_Diagnosis_ABD PAIN  Other_Diagnosis_ABD PX  \
CPSC_Case_Number                                                     
180125260                                0                       0   
180108428                                0                       0   
180120413                                0                       0   
180125238                                0                       0   
180135290                                0                       0   

                  Other_Diagnosis_ABDOMINAL PAIN  Other_Diagnosis_ABSCESS  \
CPSC_Case_Number                                                            
180125260                                      0                        0   
180108428                                      0                        0   
180120413                                      0                        0   
180125238                                      0                        0   
180135290                                      0                        0   

                  Other_Diagnosis_ACHE  Other_Diagnosis_ACUTE ABD PAIN  ...  \
CPSC_Case_Number                                                        ...   
180125260                            0                               0  ...   
180108428                            0                               0  ...   
180120413                            0                               0  ...   
180125238                            0                               0  ...   
180135290                            0                               0  ...   

                  PSU_95  PSU_96  PSU_97  PSU_98  PSU_99  PSU_100  PSU_101  \
CPSC_Case_Number                                                             
180125260              0       0       0       0       0        0        0   
180108428              0       0       0       0       0        0        0   
180120413              0       0       0       0       0        0        0   
180125238              0       0       0       0       0        0        0   
180135290              0       0       0       0       0        0        0   

                  Treatment_Day  Month_Day  Year_Day  
CPSC_Case_Number                                      
180125260                    31         12      2017  
180108428                    31         12      2017  
180120413                    31         12      2017  
180125238                    30         12      2017  
180135290                    30         12      2017  

[5 rows x 434 columns]

We see that our columsn are now converted to 0 or 1 for each category and has a 1 if it has that specific value. We dropped the first column of each category so there is no duplicate data. Now we can setup our training and testing sets. 

## Train Test Split

X will be all the columns expect the one we want to predict which to start will be "Diagnosis". The variable y will be the "Diagnosis" column that has all the different Diagnosis available as separate columns and are denoted as a 0 or a 1. We will then split our X and y values into a train (80%) and test (20%) set.

In [74]:
#Setup X and y variables
#X is the rest of the dataframe
#y is the variable to predict
X = d_data.drop("Diagnosis", axis=1)
y = pd.get_dummies(d_data['Diagnosis'])

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, test_size=0.2)
#Nothing to scale since age is the only real numerical value we are using. 

Since almost all of our data are categorical, there isn't anything to scale on the numerical side as Age is our only real numerical value. Let's now confirm the shape of each train and test value. 

In [75]:
X_train.shape

(5267, 433)

In [76]:
X_test.shape

(1317, 433)

In [77]:
y_train.shape

(5267, 27)

In [78]:
y_test.shape

(1317, 27)

We see that there are 5267 rows for the training data and 1317 rows for the test data. We also see that the X values have 433 columns while the y values have 27 columns. Let's view the X_train data to make sure it looks as we would expect it to. 

In [79]:
X_train

Age   Weight  Sex_M  Other_Diagnosis_ABD PAIN  \
CPSC_Case_Number                                                  
140248355           6  81.0979      0                         0   
131033338           3  78.8451      1                         0   
141120816           1  74.3851      1                         0   
150717427           9  15.7762      1                         0   
130910795           5   6.6878      0                         0   
...               ...      ...    ...                       ...   
150604252           2   5.6748      0                         0   
140443952           3   5.7174      0                         0   
140426349          54  14.3089      0                         0   
140125949          11  80.0213      0                         0   
170655992           4   4.7570      1                         0   

                  Other_Diagnosis_ABD PX  Other_Diagnosis_ABDOMINAL PAIN  \
CPSC_Case_Number                                                           
140248355                              0                               0   
131033338                              0                               0   
141120816                              0                               0   
150717427                              0                               0   
130910795                              0                               0   
...                                  ...                             ...   
150604252                              0                               0   
140443952                              0                               0   
140426349                              0                               0   
140125949                              0                               0   
170655992                              0                               0   

                  Other_Diagnosis_ABSCESS  Other_Diagnosis_ACHE  \
CPSC_Case_Number                                                  
140248355                               0                     0   
131033338                               0                     0   
141120816                               0                     0   
150717427                               0                     0   
130910795                               0                     0   
...                                   ...                   ...   
150604252                               0                     0   
140443952                               0                     0   
140426349                               0                     0   
140125949                               0                     0   
170655992                               0                     0   

                  Other_Diagnosis_ACUTE ABD PAIN  \
CPSC_Case_Number                                   
140248355                                      0   
131033338                                      0   
141120816                                      0   
150717427                                      0   
130910795                                      0   
...                                          ...   
150604252                                      0   
140443952                                      0   
140426349                                      0   
140125949                                      0   
170655992                                      0   

                  Other_Diagnosis_ACUTE COSTOCOMNDRISI  ...  PSU_95  PSU_96  \
CPSC_Case_Number                                        ...                   
140248355                                            0  ...       0       0   
131033338                                            0  ...       0       0   
141120816                                            0  ...       0       0   
150717427                                            0  ...       0       0   
130910795                                            0  ...       0       0   
...                                                ... 

We can see for each case number in the training data as well as their values. Now let's look at the y_train value. 

In [80]:
diagnosis = y_train.columns #Saving the column values so we can identify each column later on. 
y_train

Amputation  Anoxia  Aspiration  Avulsion  Burn, chemical  \
CPSC_Case_Number                                                             
140248355                  0       0           0         0               0   
131033338                  0       0           0         0               0   
141120816                  0       0           0         0               0   
150717427                  0       0           0         0               0   
130910795                  0       0           0         0               0   
...                      ...     ...         ...       ...             ...   
150604252                  0       0           0         0               0   
140443952                  0       0           0         0               0   
140426349                  0       0           0         0               0   
140125949                  0       0           0         0               0   
170655992                  0       0           0         0               0   

                  Burn, thermal  Concussion  Contusion, abrasion  Crushing  \
CPSC_Case_Number                                                             
140248355                     0           0                    0         0   
131033338                     0           0                    0         0   
141120816                     0           0                    0         0   
150717427                     0           0                    1         0   
130910795                     0           0                    0         0   
...                         ...         ...                  ...       ...   
150604252                     0           0                    1         0   
140443952                     0           0                    0         0   
140426349                     0           0                    0         0   
140125949                     0           0                    0         0   
170655992                     0           0                    0         0   

                  Dental injury  ...  Ingestion  Internal injury  Laceration  \
CPSC_Case_Number                 ...                                           
140248355                     0  ...          0                0           0   
131033338                     0  ...          0                0           0   
141120816                     0  ...          0                0           0   
150717427                     0  ...          0                0           0   
130910795                     0  ...          0                0           0   
...                         ...  ...        ...              ...         ...   
150604252                     0  ...          0                0           0   
140443952                     0  ...          0                0           0   
140426349                     0  ...          0                0           0   
140125949                     0  ...          0                1           0   
170655992                     0  ...          0                0           0   

                  Nerve damage  Other  Poisoning  Puncture  Radiation  \
CPSC_Case_Number                                                        
140248355                    0      0          0         0          0   
131033338                    0      0          0         0          0   
141120816                    0      1          0         0          0   
150717427                    0      0          0         0          0   
130910795                    0      0          0         0          0   
...                        ...    ...        ...       ...        ...   
150604252                    0      0          0         0          0   
140443952                    0      0          0         0          0   
140426349                    0      0          0         0          0   
140125949                    0      0          0         0          0   
170655992                    0      0          0         0          0  

We see each case number and the diagnosis they ended up with. This will give our model the training data it needs to see how the other values correlate to the diagnosis. 

## Guessing with Mean

Before we create a model, let's get a baseline down of what we would expect out of each Diagnosis by taking the mean value of the training set and get the percentage of what each Diagnosis would be. 

In [81]:
train_percent = np.mean(y_train)*100
train_percent

Amputation                     0.113917
Anoxia                         0.037972
Aspiration                     0.018986
Avulsion                       0.341751
Burn, chemical                 0.018986
Burn, thermal                  0.550598
Concussion                     3.474464
Contusion, abrasion           15.872413
Crushing                       0.379723
Dental injury                  1.082210
Dermatitis, conjunctivitis     0.512626
Dislocation                    2.050503
Electric shock                 0.037972
Foreign body                   0.265806
Fracture                      19.176002
Hematoma                       0.664515
Hemorrhage                     0.151889
Ingestion                      0.018986
Internal injury                8.239985
Laceration                    10.347446
Nerve damage                   0.303778
Other                         14.752231
Poisoning                      0.018986
Puncture                       0.208848
Radiation                      0.056958


We see that the "Strain, sprain" diagnosis makes up on average 21.2% of the cases in the training set. This shows the breakdown of what we would expect out of each Diagnosis given the training data. 

## Modeling

Now let's test a model. We can try Linear Regression first to make sure we can get the machine learning running smoothly and try to make sense of the results. 

In [83]:
#Try linear regression first

linreg = LinearRegression()
linreg.fit(X_train,y_train)

y_pred = linreg.predict(X_test)
(pd.DataFrame(y_pred, columns= diagnosis, index=y_test.index) *100).T

CPSC_Case_Number            170827787  160852455  170157166  141223560  \
Amputation                  -0.304940   0.123795   0.010496  -0.464502   
Anoxia                      -0.138998  -0.008860   0.802048  -0.757487   
Aspiration                  -0.056878  -0.015776   0.327169   0.091346   
Avulsion                     0.342918   0.352372  -0.105274  -0.506860   
Burn, chemical              -0.021937   0.103399   0.020405  -0.039538   
Burn, thermal               -0.058401  -0.781174  -0.886972  -0.452991   
Concussion                   1.288284  -1.014004  18.611813  -0.837432   
Contusion, abrasion         14.872930  32.532088   7.066135  38.753820   
Crushing                     1.385972  -0.186659   3.084086   0.741477   
Dental injury               -0.521808  -0.441588  -2.200362  -0.262640   
Dermatitis, conjunctivitis   1.559302   2.007816  -0.112130   7.377506   
Dislocation                  0.803033  -1.229095   4.828015   0.915998   
Electric shock              -0.032919  -0.122683   0.134204  -0.778669   
Foreign body                -0.098576   0.904179   0.416761  -0.203922   
Fracture                    63.430291   5.685429  -2.597440  63.089154   
Hematoma                     0.133745   2.023829   1.501114  -0.190607   
Hemorrhage                   1.391765   1.166187  -0.040269  -0.012988   
Ingestion                    0.056878   0.015776  -0.327169  -0.091346   
Internal injury              1.255364   1.012373  57.114985  -1.446494   
Laceration                  -5.096998  61.251262  15.278996  -4.501688   
Nerve damage                 0.198298   0.058204   2.393126  -0.006501   
Other                       -0.041852  -0.083557   1.022103  -0.071290   
Poisoning                    0.038443  -0.053809  -0.079308  -0.310995   
Puncture                     0.519524  -0.425458   0.000335   0.289827   
Radiation                    0.509430   0.202405  -0.302964  -0.163317   
Strain, sprain              17.911612  -3.753421  -5.806620   1.839184   
Submersion                   0.675517   0.676970  -0.153282  -1.999045   

CPSC_Case_Number            130852418  130850294  130622327  140644499  \
Amputation                  -1.350684   1.834393  -0.452217  -0.109540   
Anoxia                      -0.080827   0.072271  -0.113203   0.075700   
Aspiration                   0.066439   0.068145   0.151804   0.022288   
Avulsion                    -0.919458   3.030200  -0.568820  -1.238939   
Burn, chemical              -0.054083  -0.025689   0.057140  -0.036958   
Burn, thermal               -2.760312   1.484586   5.006734  -1.466008   
Concussion                  -0.608096  -0.499364   5.884704  -3.164835   
Contusion, abrasion         10.443583   9.708504  43.142769  -3.271033   
Crushing                    -0.576957   1.379316  -0.262002   0.637653   
Dental injury               -1.599728  -0.835333  -2.170608  -1.675106   
Dermatitis, conjunctivitis  -1.410837  -1.420155  -1.089109   2.243550   
Dislocation                 16.677478   1.945249   8.563999   1.594748   
Electric shock              -0.037667   0.021307  -0.018768   0.139240   
Foreign body                -0.769520  -0.358258  -0.062985  -0.431729   
Fracture                    28.770903  46.038683  -3.035357   4.196641   
Hematoma                    -2.104592   0.415185  -0.052418   3.190855   
Hemorrhage                  -1.314426  -0.489418  -0.298903   0.206476   
Ingestion                   -0.066439  -0.068145  -0.151804  -0.022288   
Internal injury              0.907280  -0.448551  -0.549214   0.226810   
Laceration                  20.812906  11.317666   7.272553  -6.501309   
Nerve damage                -0.479033  -0.311300   0.095769   3.415264   
Other                        0.212472   0.020619   0.150593  99.976723   
Poisoning                   -0.081112  -0.011026  -0.091504  -0.401168   
Puncture                     0.269459  -0.040061   0.423716  -0.362385   
Radiation                   -0.092500  -0.115764  -0.049535  -0.134247   
Str

Based on the results of our prediction, we can see for each case number, the expected percentage of the diagnosis. The higher the number, the more likely that is the Diagnosis for that case. 

## Conclusion

## To Do

Table of Contents

Conclusion